### PR TITLE
Feature/stale while revalidate

### PR DESCRIPTION
**Pull request by tijlleenders**
@GithinjiHans @Tushar-4781 

Gave the offline-first (stale-while-revalidate) caching for offline mode a go.
TLDR: Local is working as well as the manually deployed zinzen.me :)

image

Will need some help to make this PR clean and configure (pre-)caching details and all - but I think the basics are here.

I used the create-react-app PWA template as a basis for the registerSW and service-worker files.
This is I believe also the basis @mallikarai05 used for initiating this codebase.

I first tried withoug the Vite-PWA plugin...
The build/bundle process makes things more complicated as now we have to instruct it to also copy the service worker file - but not add a hash to it? Or maybe we do want the hash for that too? Anyway - I couldn't configure it right even though it is quite [well documented](https://rollupjs.org/guide/en/). 
This is the downside of using a builder/bundler: it creates another tool to learn and you can't easily relate source files to deployed files.
I left the rollup code commented out. If it is glaringly obvious what I'm doing wrong please teach me :)
image

Then I tried using the Vite-PWA plugin which nearly worked immediately.

So we now need to choose to go the VitePWA route or stick with Rollup to avoid yet another dev dependency/plugin.

### CODE REVIEW

**Review by pulkitvyas08 — CHANGES_REQUESTED** [review]
VitePWA seems to be a good choice. @Tushar-4781 can you look into fixing the cypress errors for this PR and maybe extend this work to add offline installation enabled for desktop, even if that doesn't get merged with this PR we will get an idea if this setup is useful or not

**Comment by tijlleenders** [comment]
Good idea.

I just checked for mobile and there it seems to be installing. We might want to make a separate issue to catch the install event so we can couple it with a button in the menu.

**Comment by pulkitvyas08** [comment]
> 

But you are testing on [zinzen.me](https://zinzen.me/). For this PR's preview URL I'm getting this error `Error during service worker registration: TypeError: ServiceWorker script at https://zinzen-git-feature-stale-while-revalidate-tijlleenders.vercel.app/service-worker.js for scope https://zinzen-git-feature-stale-while-revalidate-tijlleenders.vercel.app/ threw an exception during script evaluation.`

**Comment by tijlleenders** [comment]
There is something I removed that might be able to solve this. 

The two base files you get when using create React app (CRA) with template PWA for Typescript mention a public URL env variable - but I removed that since it was undefined when running on localhost.

**Comment by tijlleenders** [comment]
Oh wait - they're identical already.
Hmmm... that could mean trouble.

**Comment by tijlleenders** [comment]
@pulkitvyas08 I get the same error when building on localhost:
```
Uncaught non-precached-url: non-precached-url :: [{"url":"undefined/index.html"}]
```

I had solved it by adding this to the Vite-PWA plugin - but forgot to commit the fix:
```
      workbox: {
        cleanupOutdatedCaches: true, //default is true but added to make it explicit
        ignoreURLParametersMatching: []
      },
```
Committed it now.

image

**Comment by Tushar-4781** [comment]
@tijlleenders where is the **manifest** file

**Comment by tijlleenders** [comment]
> @tijlleenders where is the **manifest** file

It is being generated by the Vite-PWA plugin (hardcoded in the config - could be moved out if you prefer).
I removed the separate file because we should only have the config in one place.

**Comment by tijlleenders** [comment]
image

Works on Vercel now too. It was missing the icons.

@Tushar-4781 Could you finish up the test case fix?
Then I think we can merge.

**Review by Tushar-4781 — APPROVED** [review]
this will be the **first biggest achievement** in the **history of the ZinZen App** Cheers 🍻

**Review by pulkitvyas08 — APPROVED** [review]
This is amazing work! Installation doesn't work for firefox but does for chromium based browsers so good enough. Approved and merging 🚀

**Comment by tijlleenders** [comment]
Closes #231 #207 #208
